src/Databases/DatabaseReplicated.cpp

```diff
@@ -0,0 +1,388 @@
+{
+    return Cluster::Address::toString(getFQDNOrHostName(), global_context.getTCPPort());
+}
+
+
+DatabaseReplicated::~DatabaseReplicated() = default;
+
+DatabaseReplicated::DatabaseReplicated(
+    const String & name_,
+    const String & metadata_path_,
+    UUID uuid,
+    const String & zookeeper_path_,
+    const String & shard_name_,
+    const String & replica_name_,
+    const Context & context_)
+    : DatabaseAtomic(name_, metadata_path_, uuid, "DatabaseReplicated (" + name_ + ")", context_)
+    , zookeeper_path(zookeeper_path_)
+    , shard_name(shard_name_)
+    , replica_name(replica_name_)
+{
+    if (zookeeper_path.empty() || shard_name.empty() || replica_name.empty())
+        throw Exception("ZooKeeper path, shard and replica names must be non-empty", ErrorCodes::BAD_ARGUMENTS);
+    if (shard_name.find('/') != std::string::npos || replica_name.find('/') != std::string::npos)
+        throw Exception("Shard and replica names should not contain '/'", ErrorCodes::BAD_ARGUMENTS);
+
+    if (zookeeper_path.back() == '/')
+        zookeeper_path.resize(zookeeper_path.size() - 1);
+
+    /// If zookeeper chroot prefix is used, path should start with '/', because chroot concatenates without it.
+    if (zookeeper_path.front() != '/')
+        zookeeper_path = "/" + zookeeper_path;
+
+    if (!context_.hasZooKeeper())
+    {
+        throw Exception("Can't create replicated database without ZooKeeper", ErrorCodes::NO_ZOOKEEPER);
+    }
+    //FIXME it will fail on startup if zk is not available
+
+    auto current_zookeeper = global_context.getZooKeeper();
+
+    if (!current_zookeeper->exists(zookeeper_path))
+    {
+        /// Create new database, multiple nodes can execute it concurrently
+        createDatabaseNodesInZooKeeper(current_zookeeper);
+    }
+
+    replica_path = zookeeper_path + "/replicas/" + shard_name + "/" + replica_name;
+
+    String replica_host_id;
+    if (current_zookeeper->tryGet(replica_path, replica_host_id))
+    {
+        String host_id = getHostID(global_context);
+        if (replica_host_id != host_id)
+            throw Exception(ErrorCodes::REPLICA_IS_ALREADY_EXIST,
+                            "Replica {} of shard {} of replicated database at {} already exists. Replica host ID: '{}', current host ID: '{}'",
+                            replica_name, shard_name, zookeeper_path, replica_host_id, host_id);
+
+        log_entry_to_execute = parse<UInt32>(current_zookeeper->get(replica_path + "/log_ptr"));
+    }
+    else
+    {
+        /// Throws if replica with the same name was created concurrently
+        createReplicaNodesInZooKeeper(current_zookeeper);
+    }
+}
+
+bool DatabaseReplicated::createDatabaseNodesInZooKeeper(const zkutil::ZooKeeperPtr & current_zookeeper)
+{
+    current_zookeeper->createAncestors(zookeeper_path);
+
+    Coordination::Requests ops;
+    ops.emplace_back(zkutil::makeCreateRequest(zookeeper_path, "", zkutil::CreateMode::Persistent));
```

Comment:
Node1 : CREATE TABLE replicated_database.table_name_1(`key1` String, `value1` String) ENGINE = Log;
Node2 : CREATE TABLE replicated_database.table_name_1(`key2` String, `value2` String) ENGINE = Log;

What happens when they execute at the same time?

Reply:
Queries are ordered through replication queue in ZooKeeper, the second one will fail with "table already exists"